### PR TITLE
Add a calendar button to every DateBox

### PR DESCRIPTION
**Pull request by humdingerb**
Showing the calendar window and setting the date on the DateBox is all handled by the CalendarButton.

### CODE REVIEW

**Comment by humdingerb** [comment]
Since I can't make it work to just add a BButton to the DateBox directly...
I think it's OK since it's all pretty contained in the CalendarButton.
Tell me what you think, @dospuntos and @korli.

File: src/CheckView.cpp [hunk 1]
```diff
@@ -7,6 +8,7 @@
 #include <stdlib.h>
 
 #include "Account.h"
+#include "CalendarButton.h"
```
Comment (korli) [inline src/CheckView.cpp:11]: CalendarButton.h is missing it seems

Comment (humdingerb) [reply src/CheckView.cpp:11]: Blast! Once more I forgot to git add the main character of the show...
PR amended.

**Comment by dospuntos** [comment]
I think it looks great, and I like that the date can be edited manually as well.

Only thing I've noticed is that I don't get an icon on the calendar button, just a square (missing font?)

calendar

**Comment by humdingerb** [comment]
Hmm... I thought Haiku now supported fall-back if a font doesn't provide a symbol. It is i the standard Noto fonts though. It's [this symbol: 📅](https://www.compart.com/en/unicode/U+1F4C5)
What font are you using?

**Comment by dospuntos** [comment]
Strange, I'm also using Noto (default settings), but the symbol isn't rendered at all (not in other applications either), so it's not a problem with CapitalBe specifically. Not sure why, though... Is there anything else I can do to test this?

image

**Comment by humdingerb** [comment]
No idea. You're not on an old Beta4, are you? I think the fallback mechanism was added/improved later...

**Comment by dospuntos** [comment]
No, I'm using the nightlies, currently hrev57753.

**Comment by humdingerb** [comment]
Weird. I suspected some other insalled font on my system would maybe provide the symbol. So I uninstalled all other fonts, but the calendar icon is still there... Does your noto font package include "symbol" fonts? It should...

**Comment by humdingerb** [comment]
I've booted into a 32bit Haiku which has no extra fonts installed AFAIK, a bit older nightly, I think.
As an aside: CapitalBe crashes when built with gcc13 there, works with gcc2. Something Translator related, possibly (?). I dunno if we care at this point...

Anyway, there I too see the box instead of the symbol. I did find a similar symbol though: 🗓

another-symbol

It's just not as pretty. :)

Alternatively, you could check out this font, which I expect to have that symbol:

[NerdNotoFont.zip](https://github.com/user-attachments/files/15816075/NerdNotoFont.zip) [1.2 MiB]

If it shows the symbol after installing it, we could decide to have the CapitalBe package depend on it.

Or...

We could make an icon button of it and include the symbol as vector icon.

**Comment by humdingerb** [comment]
Here would be the version with a vector icon:

calendar-icon

It may not be quite as crisp with large or very small font sizes, but it would avoid relying on a font with the calendar character being present...
Tell me what you think, @dospuntos

**Comment by dospuntos** [comment]
I think it looks good, and it's clear what the button does. Don't know what's going on with the fonts on my system, I've installed a bunch from HaikuDepot, but I still don't see most of the symbols (🗓 does show up correctly). But the vector icon is probably a safer option for now.

**Comment by humdingerb** [comment]
The one other concern I have is that putting an icon there instead of text is not custom-colour aware. We could provide a black and a white version, but if the user sets another colour, it'll stand out.

I think I'll try - for the very first time - to create a BPicture for the button. Then we can use B_CONTROL_TEXT_COLOR and are safe. Fingers crossed...

**Comment by lonemadmax** [comment]
> Strange, I'm also using Noto (default settings), but the symbol isn't rendered at all (not in other applications either), so it's not a problem with CapitalBe specifically. Not sure why, though... Is there anything else I can do to test this?
> 
Install noto_emoji. It has the glyph and is used as fallback if found, but it is not installed by default.

**Comment by dospuntos** [comment]
> Install noto_emoji

Yes, that fixed it, now the icon is working in CapitalBe. Thanks.

@humdingerb - would i make sense to have CapitalBe depend on that font? Your ideas with BPicture/BButton/BViews sound interesting too, though, looking forward to see how it turns out.

**Comment by humdingerb** [comment]
I'm still exprimenting...
If all fails, in the worst case people installing CapitalBe will get to install the 4 MB emoji font. Not massive as dependencies go, but better avoided.

**Comment by lonemadmax** [comment]
With those little pictures used everywhere nowadays I would (wrongly, as shown here) expect everyone to have it installed already. Anyway, if you are feeling really adventurous and don't have enough work on your hands already, you can subtype it: use a font editor to extract the character(s) and features that you need or remove the ones you don't and create another font with a unique name, include it with the app and [load it just for the app's use](https://git.haiku-os.org/haiku/tree/headers/os/interface/Font.h#n285).

**Comment by humdingerb** [comment]
A good idea. However, I'm now nearing success in drawing a simple version of the icon directly in a BView. I'm astonishingly bad at the needed pixel math to align stuff... :)

**Comment by humdingerb** [comment]
OK, I finally think it's font independent enough to look OK with different sizes:

finally

I've updated the PR. Let me know of any improvements.

**Comment by lonemadmax** [comment]
And with a dynamic date, nice.

In case you need it again, you don't even have to resort to a font editor: in a dummy program set an appropriate font, call `BFont::GetGlyphShapes` and archive the BShape. In the real app, instantiate the shape and `BView::FillShape` it.

**Comment by humdingerb** [comment]
Thanks. Many roads lead to rome. Though, if I had to take that one, it'd need much experimenting for me... :)

I think, we should bundle NotoEmojy with Haiku. If we have over 100 MiB for the CJK font, we could have those measly 4 MiB emoji font...

**Comment by humdingerb** [comment]
I made a little change: The calendar icon now always just shows the current date.

I found it a bit confusing that
* you expect all widgets resulting in showing the calendar window to look the same. In case of the Report, though, you have different days for start and end dates. Looks weird...
* It's incosistent. The icon only gets updated if you set a new date with the calendar window. It won't change when editing the date box directly. This could be 'fixed', but I don't think this little gimmick justifies the added complexity.

If I don't hear objections, I'll merge this soon.

**Comment by dospuntos** [comment]
No objections on my end, this looks great. At first I expected the icon to update when selecting a date, but I agree with your notes above.

**Comment by humdingerb** [comment]
It did work like that before I noticed the mentioned drawbacks...
Let's merge then.

**Pull request merged**